widgets/heatmap/heatmap.go

```diff
@@ -89,42 +144,161 @@ func (hp *HeatMap) ClearYLabels() {
+// that can fit into the canvas.
+//
 // This is essentially the number of available cells on the canvas as observed
 // on the last call to draw. Returns zero if draw wasn't called.
 //
 // Note that this capacity changes each time the terminal resizes, so there is
 // no guarantee this remains the same next time Draw is called.
 // Should be used as a hint only.
-func (hp *HeatMap) ValueCapacity() int {
-	return 0
+func (hp *HeatMap) ValueCapacity() (rows, cols int) {
+	hp.mu.Lock()
+	defer hp.mu.Unlock()
+
+	if hp.lastWidth == 0 || hp.lastHeight == 0 {
+		return 0, 0
+	}
+
+	rows = hp.lastHeight - 1
+	var cw int
+
+	if hp.opts.cellWidth > minCellWidth {
+		cw = hp.opts.cellWidth
+	} else {
+		cw = minCellWidth
+	}
+	cols = int(math.Floor(float64(hp.lastWidth-axes.LongestString(hp.yLabels)-axes.AxisWidth) / float64(cw)))
+	return
 }
 
 // axesDetails determines the details about the X and Y axes.
 func (hp *HeatMap) axesDetails(cvs *canvas.Canvas) (*axes.XDetails, *axes.YDetails, error) {
-	return nil, nil, errors.New("not implemented")
+	hp.cellWidthAdaptive(cvs)
+
+	yd, err := axes.NewYDetails(hp.yLabels)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	xd, err := axes.NewXDetails(yd.End, hp.xLabels, hp.opts.cellWidth)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return xd, yd, nil
 }
 
 // Draw draws cells, X labels and Y labels as HeatMap.
 // Implements widgetapi.Widget.Draw.
 func (hp *HeatMap) Draw(cvs *canvas.Canvas, meta *widgetapi.Meta) error {
-	return errors.New("not implemented")
+	hp.mu.Lock()
+	defer hp.mu.Unlock()
+
+	hp.lastWidth = cvs.Area().Dx()
+	hp.lastHeight = cvs.Area().Dy()
+	// Check if the canvas has enough area to draw HeatMap.
+	needAr, err := area.FromSize(hp.minSize())
+	if err != nil {
+		return err
+	}
+	if !needAr.In(cvs.Area()) {
+		return draw.ResizeNeeded(cvs)
+	}
+
+	xd, yd, err := hp.axesDetails(cvs)
+	if err != nil {
+		return err
+	}
+	err = hp.drawCells(cvs, yd)
+	if err != nil {
+		return err
+	}
+
+	return hp.drawLabels(cvs, xd, yd)
 }
 
 // drawCells draws m*n cells (rectangles) representing the stored values.
 // The height of each cell is 1 and the default width is 3.
-func (hp *HeatMap) drawCells(cvs *canvas.Canvas, xd *axes.XDetails, yd *axes.YDetails) error {
-	return errors.New("not implemented")
+func (hp *HeatMap) drawCells(cvs *canvas.Canvas, yd *axes.YDetails) error {
+	for i := 0; i < len(hp.values); i++ {
+		for j := 0; j < len(hp.values[0]); j++ {
+			startX := yd.Start.X + axes.AxisWidth + j*hp.opts.cellWidth
+			startY := yd.Labels[i].Pos.Y
+
+			endX := startX + hp.opts.cellWidth
+			endY := startY + 1
+
+			rect := image.Rect(startX, startY, endX, endY)
+			color := hp.getCellColor(hp.values[i][j])
+
+			if err := cvs.SetAreaCells(rect, hp.opts.cellChar, cell.BgColor(color)); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
 }
 
 // drawAxes draws X labels (under the cells) and Y Labels (on the left side of the cell).
```

Comment:
Sorry, missed this before, the comment here should say `drawLabels` instead of `drawAxes`.